lti_consumer/lti_xblock.py

```diff
@@ -1100,172 +1099,70 @@ def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argu
-                str(self.location),  # pylint: disable=no-member
-                self.external_user_id,
-                exc,
-            )
-            template = loader.render_django_template('/templates/html/lti_1p3_permission_error.html', context)
-            return Response(template, status=403, content_type='text/html')
+        # Backward compatibility - copy the config to the DB
+        self._save_lti_1p3_config_to_db()
+
+        from .plugin.views import launch_gate_endpoint  # pylint: disable=import-outside-toplevel
+        request.user = self.runtime.service(self, 'user').get_current_user()
+        return launch_gate_endpoint(request, suffix)
+
+    def _save_lti_1p3_config_to_db(self):
+        """
+        Save LTI 1.3 config to the database.
+        """
+        from .api import _get_lti_config  # pylint: disable=import-outside-toplevel
+        conf = _get_lti_config(block=self)
+        # Transfer the config to DB so that the django view can initialize the LTI1P3Consumer
+        # without having to load the XBlock
+        conf.lti_config = self._lti_1p3_config_as_dict()
+        conf.config_store = conf.CONFIG_ON_DB
+        conf.save()
+        return conf
+
+    def _lti_1p3_config_as_dict(self):
+        return dict(
+            lti_1p3_launch_url=self.lti_1p3_launch_url,
+            lti_1p3_oidc_url=self.lti_1p3_oidc_url,
+            lti_1p3_tool_key_mode=self.lti_1p3_tool_key_mode,
+            lti_1p3_tool_keyset_url=self.lti_1p3_tool_keyset_url,
+            lti_1p3_tool_public_key=self.lti_1p3_tool_public_key,
+            lti_1p3_enable_nrps=self.lti_1p3_enable_nrps,
+            lti_1p3_block_key=self.lti_1p3_block_key,
+            lti_advantage_deep_linking_enabled=self.lti_advantage_deep_linking_enabled,
+            lti_advantage_deep_linking_launch_url=self.lti_advantage_deep_linking_launch_url,
+            lti_advantage_ags_mode=self.lti_advantage_ags_mode,
+            weight=self.weight,
+            display_name=self.display_name,
+        )
 
     @XBlock.handler
     def lti_1p3_access_token(self, request, suffix=''):  # pylint: disable=unused-argument
         """
         XBlock handler for creating access tokens for the LTI 1.3 tool.
-
         This endpoint is only valid when a LTI 1.3 tool is being used.
-
         Returns:
             webob.response:
                 Either an access token or error message detailing the failure.
                 All responses are RFC 6749 compliant.
-
         References:
             Sucess: https://tools.ietf.org/html/rfc6749#section-4.4.3
             Failure: https://tools.ietf.org/html/rfc6749#section-5.2
         """
-        if self.lti_version != "lti_1p3":
-            return Response(status=404)
-        if request.method != "POST":
-            return Response(status=405)
+        # Update the DB with the LTI 1.3 config and fetch the config
+        conf = self._save_lti_1p3_config_to_db()
```

Comment:
All other invocations of `_save_lti_1p3_config_to_db` are gated with the `self.lti_version == "lti_1p3"` check. Don't we need to check it here too?
Or maybe it would be better to check it directly in `_save_lti_1p3_config_to_db`?

Reply:
Makes sense. This has been changed to make the check happen in the save function.